feat(web):  add geolocation utility

Johannbr opened the pull request:
The use-case : 
Some of the assets in a library contain GPS data and some do not, it could be due to multiple reasons : 

- GPS not activated
- GPS signal blocked by obstacle
- Device does not have GPS equipment (eg: DSLR)

...
I've been looking and thinking for an automatic update for assets without GPS data, and I always found many edge cases where automatic algorithms don't work. So I decided to create a manual feature in Immich/utility. Manual feature does not mean it has to be tedious 🙂

With the feature, you :

- click on an asset with GPS data
- have a look at previous and following assets
- select assets you want to update locations
- click on update location

=> the location of the source assets is copied to every selected asset

![Screen Recording Immich Geolocation Utility - 2](https://github.com/user-attachments/assets/e5389879-f56c-418f-a3d9-31e3bafe7030)

I've run :
- [x] npm run check:all
- [x] npm run format:fix

Johannbr committed:
feat(geolocation):  add geolocation utility

jrasm91 reviewed (commented):
Looks pretty good so far!

Johannbr commented on web/src/lib/components/utilities-page/geolocation/geolocation.svelte:
```diff
@@ -0,0 +1,96 @@
+      readonly={hasGps}
+    />
+
+    {#if hasGps}
+      <div class="absolute bottom-1 end-3 px-4 py-1 rounded-xl text-xs transition-colors bg-green-500">GPS</div>
+    {:else}
+      <div class="absolute bottom-1 end-3 px-4 py-1 rounded-xl text-xs transition-colors bg-red-500">No GPS</div>
+    {/if}
```
Oops, I forgot, thanks :-)

Johannbr commented on web/src/lib/modals/GeolocationUpdateConfirmationModal.svelte:
```diff
@@ -0,0 +1,42 @@
+      })}
+    </p>
+
+    <div class="text-left text-sm">
```
I didn't know this, thanks for the tip

Johannbr committed:
Merge branch 'immich-app:main' into feat/geolocation-utility

Johannbr committed:
feat(web): geolocation utility - fix code review - 1

Johannbr committed:
feat(web): geolocation utility - fix code review - 2

Johannbr committed:
Merge branch 'main' into feat/geolocation-utility

Johannbr committed:
Merge branch 'main' into feat/geolocation-utility

Johannbr committed:
Merge branch 'main' into feat/geolocation-utility

jrasm91 committed:
chore: cleanup

alextran1502 commented:
@jrasm91 Let me know when the PR is in good shape, I will work on some stylings

Johannbr commented:
Thanks for your feedback @jrasm91 

> * It is quite annoying to add 5-6 translated strings to explain how the tool works. It should either be self-evident (not require explicit instructions, or the instructions should go on the documentation website.

Good point

> * I've removed the "assets to display" logic, since I think we should have a "pagination" solution for views like this and I don't really want to add a random translation string until that UI is finalized. I think for now, just hardcoding the view to 200 assets is fine. 

I think we could go higher than 200 assets by default, why not 500 ?

> * Related, I was wondering if it would be possible to just use the timeline view for navigation but this custom thumbnail view and custom thumbnail click action. Then we could get rid of all the date management stuff and people can just browse the timeline to find their photos like they normally would.

Yes I had the same thought, it might be better if we can use the timeline. How would you like to proceed ? Do you want to merge a first version of this feature then making an improvement ? Or do you think, we need to make it work with the timeline ?

> * I added an empty placeholder, in addition to the text, when no assets were present

Isn't the "+" of the picture a bit misleading ? People will try to upload assets, no ?

> * I moved most of the buttons to "title bar"
> * I added an option to pick the location from the map, in addition to setting it by clicking on an image with GPS data
> * I found the notification about selected GPS quite annoying, so I removed it. Also, the GPS coordinates were not actually copied to the clipboard, so I found that to be misleading
> * I moved the spinning indicator to the middle of the screen instead of just having it in among the buttons.

Thanks, that's much better

s-martin commented:
> Related, I was wondering if it would be possible to just use the timeline view for navigation but this custom thumbnail view and custom thumbnail click action. Then we could get rid of all the date management stuff and people can just browse the timeline to find their photos like they normally would.

I have exactly the use case described in the PR and as a user the timeline view is the most intuitive implementation for me.

As a developer I also think it makes sense to not have an additional view, which would need to be maintained in the future.

Johannbr committed:
Merge branch 'immich-app:main' into feat/geolocation-utility

alextran1502 commented:
iirc from here https://www.manypixels.co/gallery

jrasm91 commented:
Nah, it was https://themeisle.com/illustrations/

jrasm91 committed:
chore: feedback

Johannbr commented:
Is this PR ok with you @jrasm91 ?
It seems to me that @alextran1502 wanted to work on some stylings.

alextran1502 committed:
Merge branch 'main' into feat/geolocation-utility

alextran1502 commented:
I have a few comments that I hope you can work on before getting this feature in

- When first opening the page, it'd be more useful to display instructions of how to use the tool, replacing the "No assets to show" placeholder, or a combination of "No assets to show" and the instruction
- It is not clear to me that when clicking on the assets with GPS information, it will update the selected GPS info. Perhaps there should be a different button below the card, like "Use this photo's location"
- I think the selected GPS information should be more prominent


Let me know if you can work on those. Otherwise, I can take a look at them, but probably not for a while.

jrasm91 commented:
I'd love for it to just be more obvious how to use the page. Adding instructions just means more stuff that has to be translated and more likely that it becomes outdated.

I'm just not sure what the best way to do that is.

Johannbr commented:
> When first opening the page, it'd be more useful to display instructions of how to use the tool, replacing the "No assets to show" placeholder, or a combination of "No assets to show" and the instruction

On a previous commit, there was an information icon and a modal with information. Should we revert it ?

> * It is not clear to me that when clicking on the assets with GPS information, it will update the selected GPS info. Perhaps there should be a different button below the card, like "Use this photo's location"

I added a text on the thumbnail:
<img width="396" height="315" alt="image" src="https://github.com/user-attachments/assets/2b0d5ef3-ee99-46fd-9cbc-6bcc245fa850" />

> * I think the selected GPS information should be more prominent

I added an animation on the location:
https://github.com/user-attachments/assets/a3b78d11-8e79-4084-94ca-879d2df06f78

Johannbr committed:
feat(web): add animation and text

animation on locations change and action text on thumbnail

Johannbr committed:
Merge branch 'immich-app:main' into feat/geolocation-utility

alextran1502 committed:
Merge branch 'main' into feat/geolocation-utility

alextran1502 committed:
styling, messages and filtering

alextran1502 committed:
selected color

alextran1502 committed:
format i18n

alextran1502 committed:
fix lint

pretubedongit commented:
Hey many I know if this adding GPS function is at database level only or file level? I believe it's in database level only. Can we make it optional to write the GPS info the file metadara with exiftool? With some buttons and confirmation dialogues.

AarjavPatni commented:
This doesn't work directly on the file unfortunately. Downloading the file doesn't show the updated location data.
